### PR TITLE
vim-patch:8.2.3664,8.2.3743,8.2.3747,8.2.3748,8.2.3757

### LINKED ISSUE #16494: Highlight groups not properly cleared ("Too many highlight and syntax groups")
### Neovim version (nvim -v)

NA commit 73b35ef10f955 (git describe: v0.5.0-1213-g73b35ef10)

### Vim (not Nvim) behaves the same?

no, vim 8.2.3582

### Operating system/version

Arch Linux

### Terminal name/version

kitty 0.23.1

### $TERM environment variable

tmux-256color

### Installation

Arch User Repository (neovim-git)

### How to reproduce the issue

The error is pretty generic, see the below explanation. It can be easily reproduced using the following `.vimrc` if one provokes many changes that update highlighting groups.

```vimrc
" This line should not be removed as it ensures that various options are
" properly set to work with the Vim-related packages.
runtime! archlinux.vim
" Move around code blocks in e.g. Julia
runtime macros/matchit.vim

call plug#begin('~/.vim/plugged')
Plug 'mhinz/vim-signify'  " VCS sign indicators
call plug#end()
```

### Expected behavior

No errors regardless of how long I keep my vim session open.

### Actual behavior

After a while errors similar to this one `E849: Too many highlight and syntax groups` pop up and all new highlighting in vim stops working. This can be further confirmed by looking at the highlighting groups (`:hi`) which contain thousands of the following lines:
```
           xxx cleared
           xxx cleared
           xxx cleared
           xxx cleared
           xxx cleared
```

The problem seems to be that syntax groups (here induced by vim-signify but any plugin that frequently updates highlighting groups breaks nvim) are not properly cleared and nvim eventually runs into the limit specified at [nvim/syntax.c](https://github.com/neovim/neovim/blob/73b35ef/src/nvim/syntax.c#L101). This does not happend with nvim-v0.4.3 thus I assume it is a bug in nvim rather than the various plugins that update highlighting groups.

### PR DESCRIPTION
**Pull request by jamessan**
#### vim-patch:8.2.3664: cannot adjust sign highlighting for 'cursorline'

Problem:    Cannot adjust sign highlighting for 'cursorline'.
Solution:   Add CursorLineSign and CursorLineFold highlight groups.
            (Gregory Anders, closes vim/vim#9201)
https://github.com/vim/vim/commit/e413ea04b716effb28eb49dbc98ad3f9f761545a

#### vim-patch:8.2.3743: ":sign" can add a highlight group without a name

Problem:    ":sign" can add a highlight group without a name.
Solution:   Give an error if the group name is missing. (closes vim/vim#9280)
https://github.com/vim/vim/commit/5e18ccc60bdddc4aa39ab039f1a7c918f29e67ce

#### vim-patch:8.2.3747: cannot remove highlight from an existing sign

Problem:    Cannot remove highlight from an existing sign. (James McCoy)
Solution:   Only reject empty argument for a new sign.
https://github.com/vim/vim/commit/0bac5fc5e125b7aa0f3b596c9b7f4381279e6688

#### vim-patch:8.2.3748: giving an error for an empty sign argument breaks a plugin

Problem:    Giving an error for an empty sign argument breaks a plugin.
Solution:   Do not give an error.
https://github.com/vim/vim/commit/e5710a02cb78c2a0a868ea55740835c78ddecbb4

#### vim-patch:partial 6304be625ce4

Update runtime files.
https://github.com/vim/vim/commit/6304be625ce44dcfedc6735164d0b853578581c8

Remaining changes left out of 03d250eb4504d5168a754d0f3b7e9992337d60b4

#### vim-patch:8.2.3757: an overlong highlight group name is silently truncated

Problem:    An overlong highlight group name is silently truncated.
Solution:   Give an error if the name is too long. (closes vim/vim#9289)
https://github.com/vim/vim/commit/f7f7aaf8aaad34a38d3f159e031c5bcf3394f8f1

---
Closes #16494

### CODE REVIEW

**Review by zeertzjq — COMMENTED** [review]
IIRC a change in `cursor_spec.lua` is needed every time a highlight group has been added. E.g. <https://github.com/neovim/neovim/commit/32024787b6b1ef39c23bda1653c23b5dfbde9e81#diff-d85bbab3c3d4b8241f7b2330a679e65388cd7b62f7e00f064bc49e993735374f>.

**Comment by jamessan** [comment]
> IIRC a change in `cursor_spec.lua` is needed every time a highlight group has been added. E.g. [3202478#diff-d85bbab3c3d4b8241f7b2330a679e65388cd7b62f7e00f064bc49e993735374f](https://github.com/neovim/neovim/commit/32024787b6b1ef39c23bda1653c23b5dfbde9e81#diff-d85bbab3c3d4b8241f7b2330a679e65388cd7b62f7e00f064bc49e993735374f).

If that's the case, then the test should really fail when that doesn't happen...

**Comment by zeertzjq** [comment]
There are some merge conflicts now as #16538 has been merged.

**Comment by zeertzjq** [comment]
There are merge conflicts again after #16194 was merged.

**Comment by jamessan** [comment]
I'll update again once https://github.com/vim/vim/pull/9289 is handled, so I can include those changes.

**Comment by zeertzjq** [comment]
The test introduced in <https://github.com/vim/vim/commit/f7f7aaf8aaad34a38d3f159e031c5bcf3394f8f1> seems to depend on some earlier patches.

**Comment by jamessan** [comment]
> The test introduced in [vim/vim@f7f7aaf](https://github.com/vim/vim/commit/f7f7aaf8aaad34a38d3f159e031c5bcf3394f8f1) seems to depend on some earlier patches.

I'll just add the function with that one test for now.  The rest can be fleshed out when the older patches are ported.

File: src/nvim/syntax.c [hunk 1]
```diff
@@ -7623,10 +7627,9 @@ int syn_name2id(const char *name)
 int syn_name2id_len(const char_u *name, size_t len)
   FUNC_ATTR_NONNULL_ALL
 {
-  char name_u[201];
+  char name_u[MAX_SYN_NAME + 1];
 
-  if (len == 0 || len > 200) {
-    // ID names over 200 chars don't deserve to be found!
+  if (len == 0 || len > MAX_SYN_NAME) {
```
Comment (erw7) [inline src/nvim/syntax.c:7632]: I think this condition can be removed or changed to the following `assert`. I believe this condition is no longer met by vim/vim@5e18ccc and vim/vim@f7f7aaf.

```C
assert(len != 0);
assert(len <= MAX_SYN_NAME);
```

Comment (jamessan) [reply src/nvim/syntax.c]: Actually, the API uses this path to report back the name was invalid.

**Pull request merged**